File: sdk/php/src/Internal/Infra/KcSigner.php
<?php

namespace KuCoin\UniversalSDK\Internal\Infra;

use Exception;

/**
 * KcSigner handles KuCoin request signing for both user and broker modes.
 */
class KcSigner
{
    /** @var string */
    private $apiKey;

    /** @var string */
    private $apiSecret;

    /** @var string */
    private $apiPassphrase;

    /** @var string */
    private $brokerName;

    /** @var string */
    private $brokerPartner;

    /** @var string */
    private $brokerKey;

    public function __construct(
        $apiKey = '',
        $apiSecret = '',
        $apiPassphrase = '',
        $brokerName = '',
        $brokerPartner = '',
        $brokerKey = ''
    )
    {
        $this->apiKey = $apiKey;
        $this->apiSecret = $apiSecret;
        $this->apiPassphrase = ($apiPassphrase && $apiSecret)
            ? $this->sign($apiPassphrase, $apiSecret)
            : $apiPassphrase;

        $this->brokerName = $brokerName;
        $this->brokerPartner = $brokerPartner;
        $this->brokerKey = $brokerKey;

        if (empty($apiKey) || empty($apiSecret) || empty($apiPassphrase)) {
            error_log('[AUTH WARNING] API credentials incomplete. Access is restricted to public interfaces only.');
        }
    }

    /**
     * Generate HMAC-SHA256 signature and return base64-encoded string
     *
     * @param string $plain
     * @param string $key
     * @return string
     */
    private function sign($plain, $key)
    {
        return base64_encode(hash_hmac('sha256', $plain, $key, true));
    }

    /**
     * Return headers for general signed API requests
     *
     * @param string $plain
     * @return array
     */
    public function headers($plain)
    {
        $timestamp = (string)(int)(microtime(true) * 1000);
        $signatureInput = $timestamp . $plain;
        $signature = $this->sign($signatureInput, $this->apiSecret);

        return [
            'KC-API-KEY' => $this->apiKey,
            'KC-API-PASSPHRASE' => $this->apiPassphrase,
            'KC-API-TIMESTAMP' => $timestamp,
            'KC-API-SIGN' => $signature,
            'KC-API-KEY-VERSION' => '3',
        ];
    }

    /**
     * Return broker-specific headers including partner signature
     *
     * @param string $plain
     * @return array
     * @throws Exception
     */
    public function brokerHeaders($plain)
    {
        if (empty($this->brokerPartner) || empty($this->brokerName)) {
            error_log('[BROKER ERROR] Missing broker information');
            throw new \RuntimeException('Broker information cannot be empty');
        }

        $timestamp = (string)(int)(microtime(true) * 1000);
        $signatureInput = $timestamp . $plain;
        $signature = $this->sign($signatureInput, $this->apiSecret);

        $partnerInput = $timestamp . $this->brokerPartner . $this->apiKey;
        $partnerSignature = $this->sign($partnerInput, $this->brokerKey);

        return [
            'KC-API-KEY' => $this->apiKey,
            'KC-API-PASSPHRASE' => $this->apiPassphrase,
            'KC-API-TIMESTAMP' => $timestamp,
            'KC-API-SIGN' => $signature,
            'KC-API-KEY-VERSION' => '3',
            'KC-API-PARTNER' => $this->brokerPartner,
            'KC-BROKER-NAME' => $this->brokerName,
            'KC-API-PARTNER-VERIFY' => 'true',
            'KC-API-PARTNER-SIGN' => $partnerSignature,
        ];
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/KcSigner.java
package com.kucoin.universal.sdk.internal.infra;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import lombok.extern.slf4j.Slf4j;

/** Generates KuCoin authentication headers for both normal user mode and broker mode. */
@Slf4j
public final class KcSigner {

  private static final String HMAC_SHA256 = "HmacSHA256";

  private final String apiKey;
  private final String apiSecret;
  private final String apiPassphrase;

  private final String brokerName;
  private final String brokerPartner;
  private final String brokerKey;

  public KcSigner(
      String apiKey,
      String apiSecret,
      String apiPassphrase,
      String brokerName,
      String brokerPartner,
      String brokerKey) {

    this.apiKey = nullSafe(apiKey);
    this.apiSecret = nullSafe(apiSecret);

    this.apiPassphrase =
        (!this.apiSecret.isEmpty() && !nullSafe(apiPassphrase).isEmpty())
            ? sign(apiPassphrase, this.apiSecret)
            : apiPassphrase;

    this.brokerName = nullSafe(brokerName);
    this.brokerPartner = nullSafe(brokerPartner);
    this.brokerKey = nullSafe(brokerKey);

    if (this.apiKey.isEmpty() || this.apiSecret.isEmpty() || this.apiPassphrase.isEmpty()) {
      log.warn(
          "[AUTH WARNING] API credentials incomplete. Access is restricted to public endpoints.");
    }
  }

  /** Base64-encoded HMAC-SHA256. */
  private static String sign(String plain, String key) {
    try {
      Mac mac = Mac.getInstance(HMAC_SHA256);
      mac.init(new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), HMAC_SHA256));
      byte[] raw = mac.doFinal(plain.getBytes(StandardCharsets.UTF_8));
      return java.util.Base64.getEncoder().encodeToString(raw);
    } catch (Exception e) {
      throw new IllegalStateException("HMAC-SHA256 failure", e);
    }
  }

  /** Milliseconds since epoch as String. */
  private static String ts() {
    return Long.toString(Instant.now().toEpochMilli());
  }

  /** Headers for normal signed request. */
  public Map<String, String> headers(String plain) {
    String timestamp = ts();
    String sig = sign(timestamp + plain, apiSecret);

    Map<String, String> h = new HashMap<>();
    h.put("KC-API-KEY", apiKey);
    h.put("KC-API-PASSPHRASE", apiPassphrase);
    h.put("KC-API-TIMESTAMP", timestamp);
    h.put("KC-API-SIGN", sig);
    h.put("KC-API-KEY-VERSION", "3");
    return h;
  }

  /** Headers for broker request (includes partner signature). */
  public Map<String, String> brokerHeaders(String plain) {
    if (brokerPartner.isEmpty() || brokerName.isEmpty()) {
      System.err.println("[BROKER ERROR] Missing broker information");
      throw new IllegalStateException("Broker information cannot be empty");
    }

    String timestamp = ts();
    String sig = sign(timestamp + plain, apiSecret);
    String partnerSig = sign(timestamp + brokerPartner + apiKey, brokerKey);

    Map<String, String> h = new HashMap<>();
    h.put("KC-API-KEY", apiKey);
    h.put("KC-API-PASSPHRASE", apiPassphrase);
    h.put("KC-API-TIMESTAMP", timestamp);
    h.put("KC-API-SIGN", sig);
    h.put("KC-API-KEY-VERSION", "3");
    h.put("KC-API-PARTNER", brokerPartner);
    h.put("KC-BROKER-NAME", brokerName);
    h.put("KC-API-PARTNER-VERIFY", "true");
    h.put("KC-API-PARTNER-SIGN", partnerSig);
    return h;
  }

  private static String nullSafe(String s) {
    return s == null ? "" : s.trim();
  }
}
